### ISSUE TITLE
Bad argument -> segfault

### ISSUE THREAD

**Issue by richlv** [radiator-software/p5-net-ssleay#270]
Perl v5.32.0
perl-Net-SSLeay-1.90-1.1.x86_64, OpenSUSE Tumbleweed

segfault.pl:
```
#!/usr/bin/env perl

use strict;
use warnings;
use Net::SSLeay;

my $cert = "-";
my $time = Net::SSLeay::X509_get_notAfter($cert);
```

```
> ./segfault.pl 
Argument "-" isn't numeric in subroutine entry at ./zext-segfault.pl line 8.
Segmentation fault (core dumped)
```

**Comment by h-vn** [comment]
There was a similar ticket in Net::SSLeay RT: https://rt.cpan.org/Public/Bug/Display.html?id=107432

I think in this case too we'd keep the simple wrapper approach of not trying to catch bad values within Net::SSLeay.

**Comment by gwyn17** [comment]
I do not believe that dumping core is a valid response to bad values.  SSL is integral to all Internet connections these days.   Dumping core without any indication of why or how it happened is not helpful for fixing bugs.  And provoking a core dumps is an important step in developing of exploits.

I spend a little time looking at the problem.  The core is that structure pointers such as `SSL *` or `x509 * `are mapped to `T_PTR`, where they should be `T_PTROBJ`.   A simple find/replace can fix this, and xsubpp will generate the type checking for us.

The second problem when structure pointer are passed back to perl via callbacks and such, the code uses `XPUSHs(sv_2mortal(newSViv(PTR2IV(c))));` This happens 35 places in SSLeay.xs.   At the very least,  the code should be using UVs and not IVs, from my understanding.  Ideally they would be using `sv_setref_pv(RETVALSV, "SSLPtr", (void*)RETVAL);` or equivalent.

PUSH and XPUSH are called 208 times, and all of those should be reviewed.

I'd love to work on a patch for these issues, but if the maintainers aren't interested then that would be wasted effort.  Unless I fork the code as `Net::SSLeay::Safe`, which doesn't seem like a good idea.
